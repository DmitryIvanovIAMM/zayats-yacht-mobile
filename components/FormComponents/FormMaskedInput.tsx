import { errorColor, secondary } from "@/constants/Colors";
import React, { forwardRef, useImperativeHandle, useRef } from "react";
import { Controller, useFormContext } from "react-hook-form";
import { Platform, StyleSheet, Text, TextInput, View } from "react-native";
import { MaskedTextInput, MaskedTextInputProps } from "react-native-mask-text";
import { FormInputRef } from "./FormInput";

interface FormMaskedInputProps extends MaskedTextInputProps {
  name: string;
  label: string;
  onLayoutY?: (y: number) => void; // report Y position like FormInput
  disabled?: boolean; // NEW
}

const FormMaskedInput = forwardRef<FormInputRef, FormMaskedInputProps>(
  (
    {
      name,
      label,
      style,
      onLayoutY,
      disabled,
      onChangeText: userOnChangeText,
      ...props
    },
    ref
  ) => {
    const inputRef = useRef<TextInput | null>(null);
    const containerRef = useRef<View | null>(null);
    const { control } = useFormContext();

    useImperativeHandle(ref, () => ({
      focus: () => {
        inputRef.current?.focus();
      },
    }));

    return (
      <View
        style={styles.inputGroup}
        ref={containerRef}
        onLayout={(e) => onLayoutY?.(e.nativeEvent.layout.y)}
      >
        <Text style={styles.label}>{label}</Text>
        <Controller
          control={control}
          name={name}
          render={({
            field: { onChange, onBlur, value },
            fieldState: { error },
          }) => (
            <>
              <View
                style={[
                  styles.paperLikeInput,
                  error ? styles.inputError : styles.inputBorder,
                ]}
              >
                <MaskedTextInput
                  ref={inputRef as any}
                  style={[styles.input, style]}
                  placeholderTextColor={secondary.dark}
                  onBlur={onBlur}
                  onChangeText={(text, rawText) => onChange(rawText)}
                  value={(value as any) ?? ""}
                  editable={!disabled}
                  {...props}
                />
              </View>

              <View style={styles.errorContainer}>
                {error ? (
                  <Text
                    style={styles.error}
                    numberOfLines={1}
                    ellipsizeMode="tail"
                  >
                    {error.message}
                  </Text>
                ) : null}
              </View>
            </>
          )}
        />
      </View>
    );
  }
);

export default FormMaskedInput;

const styles = StyleSheet.create({
  inputGroup: {
    marginBottom: 8, // same as in FormInput
  },
  label: Platform.select({
    default: { fontSize: 15, marginBottom: 6, color: "#222" },
    android: { fontSize: 13, marginBottom: 4, color: "#222" },
  }),
  paperLikeInput: {
    borderWidth: 1,
    borderColor: secondary.dark,
    borderRadius: 4,
    backgroundColor: "#fafafa",
    minHeight: 40, // use minHeight instead of fixed height
    paddingHorizontal: 8,
    justifyContent: "center",
  },
  input: {
    fontSize: Platform.select({ default: 16, android: 14 }) as number,
    color: secondary.dark,
    padding: 0,
    margin: 0,
  },
  inputBorder: {
    borderColor: secondary.dark,
  },
  inputError: {
    borderColor: errorColor,
  },
  errorContainer: {
    minHeight: 18,
    marginTop: 2,
    justifyContent: "center",
  },
  error: {
    color: errorColor,
    fontSize: 14,
  },
});
