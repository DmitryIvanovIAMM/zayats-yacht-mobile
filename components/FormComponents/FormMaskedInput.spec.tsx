import { act, fireEvent, render, waitFor } from "@testing-library/react-native";
import React from "react";
import { FormProvider, useForm } from "react-hook-form";
import FormMaskedInput from "./FormMaskedInput";

// Mock react-native-mask-text to a simple stub component (lint-safe)
jest.mock("react-native-mask-text", () => {
  const React = jest.requireActual<typeof import("react")>("react");
  const { View, Text, TouchableOpacity } =
    jest.requireActual<typeof import("react-native")>("react-native");

  const MaskedTextInput = React.forwardRef<any, any>(
    function MockMaskedTextInput(props, ref) {
      React.useEffect(() => {
        const target = { focus: () => {} };
        if (typeof ref === "function") ref(target);
        else if (ref) (ref as any).current = target;
      }, [ref]);

      return (
        <View accessibilityLabel="mock-masked-input" {...props}>
          {props.placeholder ? <Text>{props.placeholder}</Text> : null}
          {props.editable === false ? <Text>DISABLED</Text> : null}
          <TouchableOpacity
            accessibilityLabel="enter-masked"
            onPress={() => props.onChangeText?.("12-34", "1234")}
          />
        </View>
      );
    }
  );
  MaskedTextInput.displayName = "MockMaskedTextInput";

  return { __esModule: true, MaskedTextInput };
});

type FormValues = { phone: string };

function TestFormProvider({
  children,
  defaultValues,
  onReady
}: {
  children: React.ReactNode;
  defaultValues?: Partial<FormValues>;
  onReady?: (methods: any) => void;
}) {
  const methods = useForm<FormValues>({
    defaultValues: { phone: "", ...(defaultValues || {}) }
  });
  React.useEffect(() => {
    onReady?.(methods);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return <FormProvider {...methods}>{children}</FormProvider>;
}

function renderWithForm(
  ui: React.ReactElement,
  defaultValues?: Partial<FormValues>
) {
  let methodsRef: any;
  const utils = render(
    <TestFormProvider
      defaultValues={defaultValues}
      onReady={(m) => (methodsRef = m)}
    >
      {ui}
    </TestFormProvider>
  );
  return { ...utils, methods: methodsRef as any };
}

describe("FormMaskedInput", () => {
  it("matches snapshot", () => {
    const { toJSON } = renderWithForm(
      <FormMaskedInput
        name="phone"
        label="Phone"
        placeholder="(123) 456-7890"
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders label and placeholder, respects disabled state", () => {
    const { getByText } = renderWithForm(
      <FormMaskedInput
        name="phone"
        label="Phone"
        placeholder="(123) 456-7890"
        disabled
      />
    );
    expect(getByText("Phone")).toBeTruthy();
    expect(getByText("(123) 456-7890")).toBeTruthy();
    expect(getByText("DISABLED")).toBeTruthy();
  });

  it("stores raw value from onChangeText into form state", async () => {
    const { getByLabelText, methods } = renderWithForm(
      <FormMaskedInput
        name="phone"
        label="Phone"
        placeholder="(123) 456-7890"
      />
    );

    await act(async () => {
      fireEvent.press(getByLabelText("enter-masked"));
    });

    await waitFor(() => expect(methods.getValues("phone")).toBe("1234"));
  });

  it("shows error message when react-hook-form sets an error", async () => {
    const { getByText, methods } = renderWithForm(
      <FormMaskedInput
        name="phone"
        label="Phone"
        placeholder="(123) 456-7890"
      />
    );

    await act(async () => {
      methods.setError("phone", { message: "Phone is required" });
    });

    await waitFor(() => expect(getByText("Phone is required")).toBeTruthy());
  });
});
