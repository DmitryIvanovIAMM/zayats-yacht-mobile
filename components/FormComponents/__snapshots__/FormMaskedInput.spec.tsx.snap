// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormMaskedInput matches snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "marginBottom": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#006666",
        "fontSize": 15,
        "marginBottom": 6,
      }
    }
  >
    Phone
  </Text>
  <View
    style={
      [
        {
          "backgroundColor": "#fafafa",
          "borderColor": "#006666",
          "borderRadius": 0,
          "borderWidth": 1,
          "justifyContent": "center",
          "minHeight": 40,
          "paddingHorizontal": 8,
        },
        {
          "borderColor": "#006666",
        },
      ]
    }
  >
    <View
      accessibilityLabel="mock-masked-input"
      editable={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="(123) 456-7890"
      placeholderTextColor="#006666"
      style={
        [
          {
            "color": "#006666",
            "fontSize": 16,
            "margin": 0,
            "padding": 0,
          },
          undefined,
        ]
      }
      value=""
    >
      <Text>
        (123) 456-7890
      </Text>
      <View
        accessibilityLabel="enter-masked"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "justifyContent": "center",
        "marginTop": 2,
        "minHeight": 18,
      }
    }
  />
</View>
`;
